---
title: <img src="http://www.ub.edu/molevol/PG&ME/UB-logo.png" style="float:right;" width="40%"/></br></br>**Estimation of nucleotide divergences and evolutionary rates**
date: "May 22, 2024"
output:
  html_document:
    self_contained: yes
    theme: yeti
    font: 16px
    css: styles.css
subtitle: Computer practice 4 - Population Genetics & Molecular Evolution - Bachelor's Degree in Bioinformatics
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(comment="")
```

<head>
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.15.1/css/all.css">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.15.1/css/v4-shims.css">
  <meta name="viewport" content="width=device-width;initial-scale=1.0">
</head>

<style type="text/css">
  body{
  font-size: 12pt;
}
</style>

<div style="background-color: #86CBBB; 1px; height:3px"></div>
<br>
<div style = "background-color:#86CBBB; padding: 10px">
  <i class="fas fa-bullseye" style="font-size:25px;"></i> **Main goals of the session**
  
  1. Retrieve sequences from databases and build **multiple sequence alignments**.
  2. Estimate species **divergence** at different functional gene regions
  3. Calculate and compare evolutionary rates of two proteins with **different functions**.

</div>

***

## 1. Practicals organization

To put into practice the concepts of species divergence and evolutionary rates that you will learn in the theory lessons, you will analyse nucleotide divergence in two genes (_obp83b_ and _rpL32_) with different biological functions in five closely related species of the genus _Drosophila_.

Throughout the document you will see different icons whose meaning is:

> <p><i class="fas fa-info-circle" style="font-size:15px;"></i><font size=2> : Additional or useful information</font></p>
> <p><i class="fa fa-cogs" style="font-size:15px;"></i><font size=2>  : Practical exercise</font></p>
> <p><i class="fa fa-key" style="font-size:15px;"></i><font size=2>  : Hint to solve an exercise or to do a task</font></p> 
> <p><i class="fas fa-comment-dots" style="font-size:15px;"></i><font size=2>  : Slot to answer a question</font></p>
> <p><i class="fas fa-tasks" style="font-size:15px;"></i><font size=2>  : Problem or task to be solved</font></p>

***

## 2. Installing R packages

You can use either the `R` console in the terminal or `RStudio` for this practice. If you don't have `R` installed, you can download the appropriate package for your system [here](https://cran.rstudio.com/). To install `RStudio`, go to [this page](https://posit.co/download/rstudio-desktop/) and follow the instructions.

Before starting the exercises, you will need to install some necessary libraries for downloading sequences from the [_GenBank_](https://www.ncbi.nlm.nih.gov/genbank/) database, performing multiple sequence alignments, building phylogenetic trees, and calibrating these trees. Open the `R` console in the terminal (or in `RStudio`) and type:

<!-- replace R with {r} to execute the code when knit -->
```{r package installation, message=FALSE, warning=FALSE, results=FALSE}
packages <- c("reutils", "ape", "seqinr", "phangorn", "phylotools")
#install.packages(setdiff(packages, rownames(installed.packages())))
#install.packages(packages)
```


***

## 3. Retrieving sequence data

You can use the `efetch()` function implemented in the `reutils` package to retrieve sequences from databases. These functions allow you to connect to _GenBank_ and download the sequences directly to your computer in FASTA format. You will need the _GenBank_ identifier of either the gene region or the complete genome sequence of the species and the coordinates of the desired gene region in that genome. **Note that for some species the gene is encoded in the reverse complement sequence relative to the reference in the database. In these cases you will need to specify the strand in `efetch`**. In the example below

Let's see how to retrieve the coding regions (CDS) of the _obp83b_ gene in different species. Table 1 lists the identifiers and coordinates of this gene region in the five species:

***

**Table 1**. Identifiers and genomic coordinates of the _obp83b_ gene in different _Drosophila_ species.

<div style="overflow-x:auto;">
  <table align="left" style="width:100%;">
    <tr align="left">
      <th>Species</th>
      <th>ID</th>
      <th>Coordinates in the genome</th>
      <th>Strand</th>
    </tr>
    <tr>
      <td>_D. melanogaster_</td>
      <td> NT_033777</td>
      <td>5976177-5976817</td>
      <td>2</td>
    </tr>  
    <tr>
      <td>_D. simulans_</td>
      <td> NC_052523</td>
      <td>2470507-2471257</td>
      <td>2</td>
    </tr>
    <tr>
      <td>_D. erecta_</td>
      <td> NW_020825194</td>
      <td>26500029-26500718</td>
      <td>1</td>
    </tr>
    <tr>
      <td>_D. elegans_</td>
      <td> NW_024545863</td>
      <td>12746083-12746802</td>
      <td>2</td>
    </tr>
    <tr>
      <td>_D. pseudopobscura_</td>
      <td> NC_046679</td>
      <td>16336491-16337172</td>
      <td>1</td>
    </tr>
  </table>
</div>
</br>

***

> <i class="fas fa-info-circle" style="font-size:20px;"></i><font size="2"> GenBank_ identifiers can be obtained from the literature (articles) by starting a search with R libraries such as `rentrez`, directly from the web page of this database, using `esearch()` from `reutils`, or using other programmes and more advanced bioinformatics tools.
.</font> 

To download and write a file in FASTA format with the CDS sequence of the _obp83b_ gene in these species, you can use the `efetch()` function of `reutils`.

<!-- replace R with {r} to execute the code when knit -->
```{r download FASTA}
library(reutils)
library(ape)

## create a new working directory
dir.create("divergence")

## example for D. pseudopobscura:
efetch("NC_046679", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=1, seqstart=16336491, seqstop=16337172, outfile="./divergence/Dpse_obp83b_cds.fasta")
```

```{r my fasta downloads}
#Get D.melanogaster
efetch("NT_033777", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=2, seqstart=5976177, seqstop=5976817, outfile="./divergence/Dmel_obp83b_cds.fasta")

#Get D.simulans
efetch("NC_052523", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=2, seqstart=2470507, seqstop=2471257, outfile="./divergence/Dsim_obp83b_cds.fasta")

#Get D.erecta
efetch("NW_020825194", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=1, seqstart=26500029, seqstop=26500718, outfile="./divergence/Dere_obp83b_cds.fasta")

#Get D.elegans
efetch("NW_024545863", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=2, seqstart=12746083, seqstop=12746802, outfile="./divergence/Dele_obp83b_cds.fasta")

#Command line execution (otherwise use read.dna and write.dna)
#cat D*_obp83b_cds.fasta > obp83b_cds.fasta
EntryName<-function(folder, name){
    folder<-paste0("./", folder, "/")
    f<-paste0(folder, name,"_cds.fasta")
    sequence<-read.dna(f, format="fasta")
    seqnc<-updateLabel(sequence, labels(sequence), name)
    write.dna(seqnc, format="fasta", file=paste0(folder, "obp83b_cds.fasta"), append=TRUE)
}
new_names<-c("Dmel_obp83b", 
             "Dsim_obp83b", 
             "Dere_obp83b", 
             "Dele_obp83b", 
             "Dpse_obp83b")
for (name in new_names){
 EntryName("divergence", name) 
}
```


> <i class="fas fa-info-circle" style="font-size:20px;"></i><font size="2"> `db`, specifies the database; `rettype`, specifies the subsequence to be extracted (=CDS) and the format (=FASTA); `retmode`, specifies the file format (=plain text); `strand`, 1: positive strand, 2: negative (reverse complement with respect to the reference in the database); `seqstart` and `seqstop`, are the coordinates of the sequence to be downloaded (required if we are extracting gene regions from the complete genome sequence set).</font>

***

<div style="background-color:#FFDAA120">
<div style="background-color:#FFDAA1">
<h3>
<i class="fa fa-cogs"></i> Exercise 1
</h3>
</div>
1. Download the rest of CDS sequences of the _obp83b_ gene in the "divergence" folder and name the files as in the example (starting with a three-letter code identifying the species).
2. Create a new file (e.g. "obp83b_cds.fasta") with the CDS of the five species in fasta format (multi-fasta file).</br>
<i class="fa fa-key" style="font-size:15px;"></i> You can do this manually or, for example, use the `read.dna()` and `write.dna()` functions in the `ape` package.
</div>

***

## 4. Multiple sequence alignment

To identify the nucleotide substitutions that have occurred in the coding region of the _opb83b_ gene during the divergence of the four species, one must first predict which positions are homologous (i.e. derived from a common ancestor). There are many algorithms and programs for constructing multiple sequence alignments of DNA sequences. For purely practical reasons, in this session you will use `ape`, an `R` package that implements some of these algorithms.

<!-- replace R with {r} to execute the code when knit -->
```{r identify nucleotide substitutions}
library(ape)

## read the sequences
myseqs<-read.dna("divergence/obp83b_cds.fasta", format="fasta")

## visualize the sequences
myseqs

## multiple sequence alignment
myalign <- clustal(myseqs)

## partial visualization of the generated alignment
myalign

## write a file (maintaining the gaps introduced in the msa step)
write.dna(myalign, file="divergence/obp83b_cds_aln.fasta", format="fasta")
```
   
***

## 5. Divergence at synonymous and nonsynonymous sites

Coding regions are special for molecular evolutionary analyses because they contain two different types of positions in terms of the functional consequences of mutations, namely **synonymous** and **nonsynonymous** positions. Synonymous **changes** are those that do not alter the amino acid encoded by the codon, whereas nonsynonymous changes cause an amino acid substitution in the protein. Consequently, synonymous **sites** are sites where if a change occurs, it will be a synonymous change. The same applies to nonsynonymous sites. Studying and comparing the variability of these two types of sites is very informative about selective constraints and the functional consequences of mutations. 

Divergence is estimated as the number of substitutions **per site** (this allows the divergence of regions of different lengths to be compared). Therefore, to estimate synonymous and nonsynonymous divergence, we first need to calculate the number of sites of each type in the coding regions of interest. The most commonly used method to estimate these sites is the **Nei-Gojobori (N-G)** method (Figure 1). The N-G approach consists of estimating the proportion of changes that would be synonymous and nonsynonymous in a given codon after considering all nine possible nucleotide changes (based on the genetic code).

![](http://www.ub.edu/molevol/PG&ME/P5/N-G.png){width=45%}

***

<font size="2">**Figure 1.** Nei-Gojobori method for estimating the number of synonymous and nonsynonymous sites in a codon</font>

***

Here are some examples of functions for working with changes and synonymous sites:

<!-- replace R with {r} to execute the code when knit -->
```{r}
library(seqinr)

translate<-seqinr::translate

## function to translate a codon:
dna_to_aa <- function(codon) {
 dna<-s2c(codon)
 aa<-translate(dna, frame = 0, sens = "F", numcode = 1, NAstring = "X", ambiguous = FALSE)
 return (aa)
}  

## function to determine if a change between two codons is synonym
is_synonymous<-function(codon1, codon2) {
  return (dna_to_aa(codon1) == dna_to_aa(codon2))
}

## function to estimate the number of synonymous sites in a codon:
synpos<-function(codon) {
pos<-s2c(codon)
for (i in 1:length(pos)) {
   base = pos[i]
   BASES = c('A', 'G', 'T', 'C')
   other_bases = BASES[BASES!=base]
   syn=0
	 for (new_base in other_bases) {
		  new_pos=c(pos[0:(i-1)], new_base, pos[-(1:i)])
		  new_codon=paste(new_pos, collapse="")
	  	s<-(is_synonymous(codon, new_codon))
		  syn <- syn + length(s[s==TRUE])
	 }
   synp <- syn/3
}
return (synp)
}
```

*** 

<div style="background-color:#FFDAA120">
<div style="background-color:#FFDAA1">
<h3>
<i class="fa fa-cogs"></i> Exercise 2
</h3>
</div>
Calculate the number of synonymous and nonsynonymous **changes** and **sites** in this coding sequence alignment using the functions above:

<p style="font-family: courier new">
**Seq 1** ATGATGCAGAGTCTGTAA</br>
**Seq 2** ATGAGGCACAGTCTGTAA</br>
</p>
<i class="fa fa-key" style="font-size:15px;"></i> The number of sites of each class in each sequence is the sum over all codons in the sequence, and the total number of sites of each class in the alignment is the average over all sequences. **You can apply the functions separately for each codon, or consider a loop integrating functions such as `s2c()` and `splitseq()` from the `seqinr` package**

```{r seq calculation}
seq1<-"ATGATGCAGAGTCTGTAA"
seq2<-"ATGAGGCACAGTCTGTAA"

aa1<-dna_to_aa(seq1)
aa2<-dna_to_aa(seq2)

syn_list<-is_synonymous(seq1,seq2)
syn<-sum(syn_list)
nonsyn<-length(syn_list)-syn


changeCounter<-function(seq){
  i=3
  pre=0
  seq_Syn=0
  
  while (i<=nchar(seq)){
    s<-substr(seq, pre, i)
    
    seq_Syn=seq_Syn+synpos(s)
    pre=i+1
    i=i+3
  }
  seq_NSyn=nchar(seq)-seq_Syn
  #syn=paste0("Synonymous: ", seq_Syn)
  #nsyn=paste0("Non-synonymous: ", seq_NSyn)
  return(c(seq_Syn, seq_NSyn))
}

seq1_data<-changeCounter(seq1)
seq2_data<-changeCounter(seq2)

syn_sites<-(seq1_data[1]+seq2_data[1])/2
nsyn_sites<-(seq1_data[2]+seq2_data[2])/2
```
```{r seq conclusions, echo=FALSE}
cat("Seq1:\n\tSynonymous changes:", seq1_data[1], "\n\tNon-synonymous changes:", seq1_data[2])
cat("\nSeq2:\n\tSynonymous changes:", seq2_data[1], "\n\tNon-synonymous changes:", seq2_data[2])

cat("\n\nSynonymous sites:", syn_sites, "\nNon-synonymous sites:", nsyn_sites)

```


</div>

***

Fortunately, you will not need to repeat this calculation for all the gene sequences we will be studying in this lab. There are `R` functions that can be used to calculate the synonymous and nonsynonymous divergence of a coding region directly from the CDS alignment. The `seqinr` package implements the `KaKs()` function, which allows the calculation of these divergences. Before proceeding with the calculations, we will change the sequence names in the fasta file to make it easier to identify the species in the results matrix (now the identifiers are very long and some of them do not contain the species name):

<!-- replace R with {r} to execute the code when knit -->
```{r}
library(phylotools)

old_names<-get.fasta.name("divergence/obp83b_cds_aln.fasta")
## new names must be in the same order than old names...
new_names<-c("Dmel_obp83b", 
             "Dsim_obp83b", 
             "Dere_obp83b", 
             "Dele_obp83b", 
             "Dpse_obp83b")
ref2 <- data.frame(old_names, new_names)
rename.fasta(infile = "divergence/obp83b_cds_aln.fasta", ref_table = ref2, outfile = "divergence/obp83b_cds_renamed_aln.fasta")

cds <- read.alignment("divergence/obp83b_cds_renamed_aln.fasta",format="fasta")

## matrix with synonymous divergences (Ks)
kaks(cds)$ks

## matrix with nonsynonymous divergences (Ka)
kaks(cds)$ka
```

***

## 6. Divergence at noncoding sites

In addition to the CDS, noncoding sequences are also found in eukaryotic genes. These sequences correspond to introns, exonic untranslated regions (5' or 3') or 5' proximal regions, often containing promoters and regulatory elements. Here, you will analyse the divergence in the 5' proximal region and the introns of the obp83b gene in the same five species.

First, you will obtain noncoding sequences of this gene for all species. To do this, you need to know exon coordinates. Again, you can use `reutils` functions to read the GenBank entries in your `R` terminal:

<!-- replace R with {r} to execute the code when knit -->
```{r}
## example for D. pseudoobscura:
x<-efetch("NC_046679", db="nucleotide", rettype="gb", retmode="text", strand=1, seqstart=16336491, seqstop=16337172)
con <- content(x, as = "textConnection")
readLines(con)

#Get D.melanogaster
y<-efetch("NT_033777", db="nucleotide", rettype="gb", retmode="text", strand=2, seqstart=5976177, seqstop=5976817)
con <- content(y, as = "textConnection")
#readLines(con)

#Get D.simulans
z<-efetch("NC_052523", db="nucleotide", rettype="gb", retmode="text", strand=2, seqstart=2470507, seqstop=2471257)
con <- content(z, as = "textConnection")
#readLines(con)

#Get D.erecta
a<-efetch("NW_020825194", db="nucleotide", rettype="gb", retmode="text", strand=1, seqstart=26500029, seqstop=26500718)
con <- content(a, as = "textConnection")
#readLines(con)

#Get D.elegans
b<-efetch("NW_024545863", db="nucleotide", rettype="gb", retmode="text", strand=2, seqstart=12746083, seqstop=12746802)
con <- content(b, as = "textConnection")
#readLines(con)
```
> <i class="fas fa-info-circle" style="font-size:20px;"></i><font size="2"> The `readLines()` function prints the entry in GenBank format. Locate the feature "CDS" to find the coordinates of the coding region of the gene _obp83b_ in this species.</font>

Once you know the coordinates of CDS, you can extract 5' and intron sequences:

<!-- replace R with {r} to execute the code when knit -->
```{r}
    ## example for D.pseudobscura
    ## download the complete gene region
    efetch("NC_046679", db="nucleotide", rettype="fasta", retmode="text", strand=1, seqstart=16336491,      seqstop=16337172, outfile="divergence/Dpse_obp83b.fasta")

    ## create an object with the sequence of the complete gene region
    sequence<-read.dna("divergence/Dpse_obp83b.fasta", format="fasta")

    ## create a new sequence with only intron sequences
    ## CDS = join(54..128,186..261,317..594)
    ## 5' region = 1-53
    ## intron 1 = 129-185
    ## intron2 = 262-316
    seqnc<-sequence
    seqnc=seqnc[,c(1:53,129:185,262:316)]
    
    ## change species label
    seqnc<-updateLabel(seqnc, labels(seqnc), "Dpse_obp83b")

    ## write a fasta file with noncoding sequences of the opb83b gene of ALL SPECIES (with append=TRUE, if your run this script for all species, all sequences will be added to the same file; remember to change the ids in each case)
    write.dna(seqnc, format="fasta", file="divergence/obp83b_cds.fasta", append=TRUE)
```

Now we can align the non-coding sequences and estimate the number of substitutions per site (_K_<sub>NC</sub>) of the _obp83b_ gene. For nucleotide divergences (without any distinction of positions, just total divergence) you can use the function `dist.dna()` from the package `ape`.

<!-- replace R with {r} to execute the code when knit -->
```{r}
    ## read the noncoding sequences
    myseqs<-read.dna("divergence/obp83b_cds.fasta", format="fasta")


    ## visualize the sequences
    myseqs

    ## multiple sequence alignment
    myalign <- clustal(myseqs)

    ## partial visualization of the generated alignment
    myalign

    ## write the alignment to file
    write.dna(myalign, file="divergence/obp83b_nc_aln.fasta", format="fasta")
    
    ## read the noncoding alignment
    nc <- read.dna("divergence/obp83b_nc_aln.fasta", format="fasta")

    ## noncoding divergences in the opb83b gene
    dist.dna(nc)
```
***
<div style="background-color:#FFDAA120">
<div style="background-color:#FFDAA1">
<h3>
<i class="fa fa-cogs"></i> Exercise 3
</h3>
</div>
Fill empty cells in the following tables with the results obtained in the different divergence analyses and answer the questions:

**Table 1.** Divergence in the coding region of the gene _opb83b_(**_K_<sub>a</sub> and _K_<sub>s</sub> values are above and below the diagonal, respectively**)
```{r echo=FALSE, results='asis'}
cells <- c("**---------**","0.01992201","0.03931737","0.03749267","0.07829961",
           "0.1256942","**---------**","0.02555604","0.02529429","0.07511411",
           "0.3234201","0.3617048","**---------**","0.03576413","0.08614269",
           "0.5494927","0.7103626","0.6182696","**---------**","0.07112614",
           "0.9158393","0.9289957","1.4340955","1.0414168","**---------**")
table <- matrix(cells, nrow=5, ncol=5, byrow=T)
colnames(table) <- c("Dmel", "Dsim", "Dere", "Dele", "Dpse")
rownames(table) <- c("Dmel", "Dsim", "Dere", "Dele", "Dpse")
knitr::kable(table)

```


Divergence in the non-coding regions of the gene _opb83b_ (**Fill in the table with _K_<sub>NC</sub> values above the diagonal**)
```{r echo=FALSE, results='asis'}
cells <- c("**---------**","0.06368852","0.12667685","0.11196991","0.14909196",
           "**---------**","**---------**","0.09786105","0.12706643","0.15745580",
           "**---------**","**---------**","**---------**","0.13419579","0.22224915",
           "**---------**","**---------**","**---------**","**---------**","**---------**",
           "**---------**","**---------**","**---------**","**---------**","**---------**")
table <- matrix(cells, nrow=5, ncol=5, byrow=T)
colnames(table) <- c("Dmel", "Dsim", "Dere", "Dele", "Dpse")
rownames(table) <- c("Dmel", "Dsim", "Dere", "Dele", "Dpse")
knitr::kable(table)
```

**Questions:**

**1\. Is the _obp83b_ gene evolving rapidly? And the Obp83b protein?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r, echo=FALSE}
cat("The average values for each species are as follows: Dmel: 0.47611, Dsim: 0.66702, Dere: 1.02618, and Dele: 1.041416. These values show a significant increase, suggesting a rapid evolution of genes.
For the protein, the averages for non-synonymous substitutions are: Dmel: 0.043757, Dsim: 0.041988, Dere: 0.060953, and Dele: 0.0711261. When we compare these to the synonymous substitution averages, we observe that the non-synonymous averages are significantly lower. This indicates negative selection, suggesting that the protein is also undergoing evolutionary changes.")
```

**2\. In which of the analysed positions (synonymous, nonsynonymous or noncoding) are evolutionary distances higher? What is the reason for this, in your opinion?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r, echo=FALSE}
cat("For synonymous positions. This may be because they don't affect the function of the gene, they have no functional constraints, so more of them can happen without impacting the organism negatively.")
```

**3\. Could you consider some of the observed changes as selectively neutral? Which ones?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r, echo=FALSE}
cat("The mutations in non-coding regions can be considered as selectively neutral changes because they don't affect the evolution of genes")
```

**4\. Which species would share a more recent common ancestor? In which evolutionary distances (synonymous, nonsynonymous or noncoding) should we focus to know that?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r, echo=FALSE}
cat("We have observe the non synonymous regions for the smallest difference, the smallest difference in divergence is Dsim with Dmel with a difference in divergence of 0.0018 aprox.")
```
</div>

***

## 7. Phylogenetic tree reconstruction and tree calibration

Some of the most direct applications of nucleotide divergence estimates are the reconstruction of phylogenetic relationships and the inference of evolutionary rates. To reconstruct the phylogenetic relationships among the five _Drosophila_ species using the information form the _obp83b_ gene, you can use the functions from `ape` and `phangorn` packages.

As an example, the following code reconstructs the phylogenetic relationships between the five species based on synonymous divergences. It uses the synonymous divergence matrix estimated above and the [Neighbor Joining algorithm](ttps://academic.oup.com/mbe/article/4/4/406/1029664?login=false)

<!-- replace R with {r} to execute the code when knit -->
```{r}
library(phangorn)

cds <- read.alignment("divergence/obp83b_cds_renamed_aln.fasta",format="fasta")

## NJ tree
syntree <- NJ(kaks(cds)$ka)
plot(syntree)
write.tree(syntree, file="divergence/obp83b_Ks.tree")
```

By definition, a Neighbor Joining tree is an **unrooted tree**.  There is no root node that is the common ancestor of all species. With an unrooted tree, we can tell the phylogenetic relationships between species (i.e. which species have a more recent common ancestor compared to the other species), but not the direction of evolution (i.e. which species arose more recently). In addition, branch lengths represent substitutions per site, in this case the number of substitutions per site. Sometimes, as in our example, we know which branch contains the root of the species under study (see Figure 2).


![](http://www.ub.edu/molevol/PG&ME/P5/drosotree.png){width=250}

**Figure 2**. Phylogenetic relationships among the species of the genus _Drosophila_(modified from Mol Ecol Resour. 2022;22:1559–1581)

To calculate the evolutionary rate of the Obp83b protein, we need an ultrametic tree, i.e. a rooted tree in which the branch lengths represent time (not substitutions) and at least one calibration point (i.e. a dated node). 

<!-- replace R with {r} to execute the code when knit -->
```{r}
## read the tree to a phylo class object
tree<-read.tree(file="divergence/obp83b_Ks.tree")

## plot the unroted tree
plot(tree, main="Obp gene tree (number of nonsynonymous substitutions per site)")

## root the tree using Drosophila pseudoobscura as the outgroup
rooted<-root(tree, "Dpse_obp83b")
plot(rooted)


## set calibration
## select the node corresponding to the ancestor of D. melanogater, D. simulans and D. erecta, and     specify a range of 6.1-18.9 MYA
mrca_node <- getMRCA(rooted, c("Dmel_obp83b", "Dsim_obp83b", "Dere_obp83b")) 
cal <- data.frame(
    node = mrca_node,
    age.min = 6.1,
    age.max = 18.9
)



## fit the model (penalized likelihood)
chr<-chronos(rooted, model="clock", calibration = cal)

# plot de calibrated (ultrametric) tree
plot(chr, main="calibrated tree (million years)")
axisPhylo()
tiplabels()
nodelabels()
chr$edge
chr$edge.length
```


> <i class="fas fa-info-circle" style="font-size:20px;"></i><font size="2"> The `tiplabels()` and `nodelabels()` functions will tell you the id of each node in the tree (they are indicated in the tree). With `chr$edge` and `chr$edge.length` you can find the correspondence between edge numbers and edge lengths, and thus **the estimated date in million years of each node**.</font>

You now have all the parameters you need to calculate the evolutionary rate of the Obp83b protein in _Drosophila_.

<div style="background-color:#FFDAA120">
<div style="background-color:#FFDAA1">
<h3>
<i class="fa fa-cogs"></i> Exercise 4
</h3>
</div>
Calculate the evolutionary rate (_r_) of the Obp83b protein in _Drosophila_ using the divergences and times estimated in this practice. Think carefully about which divergence to use for this calculation and apply the formula. 
</div>
```{r}
kaks(cds)$ka
divergence_times <- c(dele_dere = 21, dele_dpse = 18, dele_dsim = 22, dele_dmel = 22)
ks_values <- c(dele_dere = 0.03576413, dele_dpse = 0.07112614, dele_dsim = 0.02529429, dele_dmel = 0.03749267)
evolutionary_rates <- ks_values /2*divergence_times
mean_evolutionary_rate <- mean(evolutionary_rates)
mean_evolutionary_rate
cat("The evolutionary rate is", mean_evolutionary_rate)
```
   
***

<div style="background-color:#faeae1">
<div style="background-color:#d1926d">
<h3>
<i class="fa fa-cogs"></i> Final assignment
</h3>
</div>

**Table 2**. Identifiers and genomic coordinates of the _rpL32_ gene in different _Drosophila_ species.

<div style="overflow-x:auto;">
  <table align="left" style="width:100%;">
    <tr align="left">
      <th>Species</th>
      <th>ID</th>
      <th>Coordinates in the genome</th>
      <th>Strand</th>
    </tr>
    <tr>
      <td>_D. melanogaster_</td>
      <td> NT_033777</td>
      <td>30045229-30046161</td>
      <td>2</td>
    </tr>  
    <tr>
      <td>_D. simulans_</td>
      <td> NC_052523</td>
      <td>26165102-26166121</td>
      <td>2</td>
    </tr>
    <tr>
      <td>_D. erecta_</td>
      <td> NW_020825194</td>
      <td>2228933..2229932</td>
      <td>1</td>
    </tr>
    <tr>
      <td>_D. serrata_</td>
      <td> NW_018367383</td>
      <td>6093667..6094458</td>
      <td>2</td>
    </tr>
    <tr>
      <td>_D. pseudopobscura_</td>
      <td> NC_046679</td>
      <td>811678-812625</td>
      <td>1</td>
    </tr>
  </table>
</div>
</br>


Using the data in Table 2, calculate the rate of evolution of the protein RpL32 in _Drosophila_ and answer the following questions. Note that for this gene _D. elegans_ is replaced by _D. serrata_.

```{r Final questions}

dir.create("Final_ass")

#RUN ONLY THE FIRST TIME
#efetch("NT_033777", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=2, seqstart= 30045229, seqstop=30046161, outfile="./Final_ass/Dmel_cds.fasta")

#efetch("NC_052523", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=2, seqstart=26165102, seqstop=26166121, outfile="./Final_ass/Dsim_cds.fasta")

#efetch("NW_020825194", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=1, seqstart=2228933, seqstop=2229932, outfile="./Final_ass/Dere_cds.fasta")

#efetch("NW_018367383", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=2, seqstart=6093667, seqstop=6094458, outfile="./Final_ass/Dser_cds.fasta")

#efetch("NC_046679", db="nucleotide", rettype="fasta_cds_na", retmode="text", strand=1, seqstart=811678, seqstop=812625, outfile="./Final_ass/Dpse_cds.fasta")

#Delete entry repetitions in Dere_obp83b_cds.fasta, Dmel_obp83b_cds, Dsim_obp83b_cds

new_names<-c("Dmel", 
             "Dsim", 
             "Dere", 
             "Dser", 
             "Dpse")
for (name in new_names){
 EntryName("Final_ass", name)
}
system("mv Final_ass/obp83b_cds.fasta Final_ass/fa_cds.fasta")
```

```{r alignment}
## read the noncoding sequences
myseqs<-read.dna("Final_ass/fa_cds.fasta", format="fasta")

## multiple sequence alignment
myalign <- clustal(myseqs)

## write the alignment to file
write.dna(myalign, file="Final_ass/fa_aln.fasta", format="fasta")
a<-read.dna("Final_ass/fa_aln.fasta", format="fasta")

old_names<-get.fasta.name("Final_ass/fa_aln.fasta")
new_names<-c("Dmel", 
              "Dsim", 
              "Dere", 
              "Dser", 
              "Dpse")
ref2 <- data.frame(old_names, new_names)
rename.fasta(infile = "Final_ass/fa_aln.fasta", ref_table = ref2, outfile = "Final_ass/fa_renamed_aln.fasta")

## read the noncoding alignment
cds <- read.alignment("Final_ass/fa_renamed_aln.fasta", format="fasta")

kaks(cds)$ks #synonymous
kaks(cds)$ka #Non-synomymous
```
```{r phylogenetic tree}
library(phangorn)

cds <- read.alignment("Final_ass/fa_renamed_aln.fasta",format="fasta")

syntree <- NJ(kaks(cds)$ks)
write.tree(syntree, file="Final_ass/fa_Ks.tree")
plot(syntree)

tree<-read.tree(file="Final_ass/fa_Ks.tree")

rooted<-root(tree, "Dere") #We choose D.erecta as outgroup because it is not in any cluster

mrca_node <- getMRCA(rooted, c("Dser", "Dpse")) #I choose this two because they form a cluster

#I use the same distances that in the example
cal <- data.frame(
    node = mrca_node,
    age.min = 6.1,
    age.max = 18.9
)
  
    chr<-chronos(rooted, model="clock", calibration = cal)

plot(chr, main="calibrated tree (million years)")
axisPhylo()
tiplabels()
nodelabels()
chr$edge
chr$edge.length
```



**5\. Which is the evolutionary rate of the RpL32 protein?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r FA_5}
divergence_times <- c(dmel_dsim = 6.1, dmel_dere = 66, dmel_dser = 66, dmel_dpse = 66)
ks_values <- c(dmel_dsim = 0.021182090, dmel_dere = 0.04469301, dmel_dser = 0.01236601, dmel_dpse = 0.01143935)
evolutionary_rates <- ks_values /2*divergence_times
mean_evolutionary_rate <- mean(evolutionary_rates)
mean_evolutionary_rate
```
```{r FA_5.1, echo=FALSE}
cat("The evolutionary rate is", mean_evolutionary_rate)
```


**6\. Are the evolutionary rates of the two proteins studied in this practice different?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r FA_6, echo=FALSE}
cat("Yes, the evolutionary rate of obp83b was 0.4265788 while this one is 0.5812629")
```

**7\. Can you provide a molecular evolutionary explanation for the answer to question 6?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r FA_7, echo=FALSE}
cat("Because of difference of presence of variables like functional constraints, selective pressures, genetic redundancy, environmental interactions, and protein essentiality which can account for the more rapid evolutionary rate of one of the 2 proteins")
```

**8\. Which of the other positions examined here (i.e. nonsynonymous, noncoding, the full cds...) could have been used to reconstruct the phylogenetic tree prior to calibration?**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r FA_8, echo=FALSE}
cat("The CDS could have been used, particularly when combined with synonymous and noncoding regions to balance the evolutionary signals.")
```

**9\. Looking at your results and the tree in Figure 2, do you think that the fact that some of the species are not the same in both analyses (_D. elegans_ in obp83b and _D. serrata_ in rpL32) influences the results? Discuss**

#### <i class="fas fa-comment-dots"></i> Answer:

```{r FA_9, echo=FALSE}
cat("The fact that they are different species probably does affect the analysis, between species there may be differences like observed evolutionary rates, divergence times, selective pressures, and lineage-specific evolutionary dynamics, which affect the analysis. We chould use the same set of species for both genes for a better analysis.")
```
</div>

***

<h1>Deliver info</h1>
<p><i class="fa fa-upload"></i><strong> Submit this document with your answers (or just a document with your answers to the questions in any readable format, e.g. word, pdf, plain text...), and the `R` code used to complete the final assignment to <a href="https://aula.esci.upf.edu/mod/assign/view.php?id=242848"> AULAESCI</a></strong></p>
<p><i class="fa fa-clock"></i> <strong>Deadline</strong>: June 28, 2024</p>
<p><i class="fa fa-envelope"></i> <strong>Doubts?</strong> <a href="mailto:mailto:alejandro.sanchez@prof.esci.upf.edu" class="email">mailto:alejandro.sanchez@prof.esci.upf.edu</a></p>

***
  
